webrtc/README.md

```diff
@@ -0,0 +1,268 @@
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Addressing
+
+WebRTC multiaddresses are composed of an IP and UDP address component, followed
+by `/webrtc` and a multihash of the certificate that the node uses.
+
+Examples:
+- `/ip4/1.2.3.4/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+- `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+
+## Connection Establishment
+
+### Browser to public Server
+
+Scenario: Browser _A_ wants to connect to server node _B_ where _B_ is publicly
+reachable but _B_ does not have a TLS certificate trusted by _A_.
+
+As a preparation browser _A_ [generates a
+certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate)
+and [gets the certificate's
+fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`),
+  through some external mechanism.
+
+2. _A_ instantiates a `RTCPeerConnection`, passing its local certificate as a
+   parameter. See
+   [`RTCPeerConnection()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection).
+
+3. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+4. _A_ establishes the connection to _B_. _B_ learns of _A_ TLS fingerprint
+   through A's STUN message USERNAME attribute. See Open Questions below for
+   potential better solutions.
+
+5. See [Connection Security](#connection-security).
+
+6. See [Multiplexing](#multiplexing).
+
+#### Open Questions
+
+- How does the server learn the TLS certificate fingerprint of the browser? Is
+  embedding A's TLS certificate fingerprint in A's STUN message USERNAME
+  attribute the best option?
+
+- Is the fact that the server accepts STUN messages from the client prone to
+  attacks?
+
+    - Can an attacker launch an **amplification attack** with the STUN endpoint
+      of the server?
+
+    - Can a client run a **DOS attack** by sending many STUN messages with
+      different ufrags using different UDP source ports, forcing the server to
+      allocate a new peer connection for each? Would rate limiting suffice to
+      defend against this attack?
```

Comment:
It seems like this is not libp2p-specific. Does the WebRTC spec say anything about this? How does Pion defend against it?

Reply:
In the case of pion, either the source IP or ufrag (in case of peer-reflexive candidates) is known before we expect any STUN messages because of the SDP and candidate exchange. Pion and webrtc-rs will check the source address and ufrag of any incoming STUN message and drop packets where it does not find any match.

Reply:
That also means if you leak your ufrag, then you can be easily DoS. But looks like it doesn't happen that often.

Reply:
Documented rate limiting in https://github.com/libp2p/specs/pull/412/commits/4de8f96c05edc8d25a26862c93ed36ea6a0e98c2. I am resolving here.